### PR TITLE
vcpkg manifest mode

### PR DESCRIPTION
**Pull request by Be-ing**
successor of #4281 using a branch on the upstream repo so GITHUB_TOKEN gets write permission

### CODE REVIEW

**Review by daschuer — CHANGES_REQUESTED** [review]
While I acknowledge that using submodule is part of the vcpkg examples, we schould verify if this is a good fit for Mixxx.

The submodul solution works probably OK for GitHub actions, but has some usability issues when used localy.
Especially if you are using git submodules to switch easily between different frature branches. A loose reference should be more convenient and become the default locally.

Once the vcpkg configuration is settled, I don't expect many changes.
I also expect, that you can use always the most recent version, even if you are working on an old feature branch.So a strict commit hash dependency is IMHO not helpfull.

CMake already checks the version of the environment at a library basis. We need to make sure this is sufficient anyway becaues we also want to build against system libraries.
If we think that is not sufficient, we may add a semantic version for the vcpkg environment.

Proposal: 
cmake may checkout the most recent vcpkg for the current development branch into, lets say a vcpkg-2.4 folder, sibiling to the Mixxx .git folder in the base working tree (the one with the real .git folder in case of git worktrees)
This way the user can have two or more vcpkg environments parallel and is able to freely create worktrees of a stable Mixxx branch or main.
There will be no hassle with submodules and the source tree will be free of files that may confuse a Linux package maintainer.

What do you think? 

We should discuss this and other options before merging this.

**Comment by Be-ing** [comment]
vcpkg is designed to be used as a submodule. I will not hack around this. Working against how tools are designed is asking for trouble.

**Comment by Be-ing** [comment]
@JoergAtGithub can you review the updated build documentation in the last commit?

**Review by JoergAtGithub — COMMENTED** [review]
In general, the Readme text is difficult to read, because it combines the build processes for all OSs with all there differences in one text.

**Comment by Be-ing** [comment]
> In general, the Readme text is difficult to read, because it combines the build processes for all OSs with all there differences in one text.

Thank you for this feedback. I have reorganized the text into subsections. Please take another look.

**Comment by Be-ing** [comment]
Thank you for the feedback. I think BUILDING.md is in good shape now. Better than our old verbose wiki pages.

**Comment by daschuer** [comment]
>> Please don't reabase a PR under review.

> Sorry, I should have marked it as draft earlier. It is now ready for review.

What is the value of saying sorry if you do it again 5 h later.  :-(

**Comment by Be-ing** [comment]
Let's merge https://github.com/mixxxdj/vcpkg/pull/26 and https://github.com/mixxxdj/vcpkg/pull/27 before merging this. Currently the submodule is using a branch on my fork of vcpkg.

**Comment by Be-ing** [comment]
Also please confirm that you have the most recent version of this branch checked out locally.

**Comment by Be-ing** [comment]
Thanks. The problem is that the CMake version is not matching GitHub Actions. I downloaded the [hidapi_x64-windows](https://github.com/mixxxdj/mixxx/packages/1046443) package (.nuget packages are just ZIP files) and compared the file share/hidapi/vcpkg_abi_info.txt with the log you posted. In the NuGet package it says CMake 3.21.3 but in your log it says CMake 3.21.1.

**Comment by Be-ing** [comment]
Potentially related links for the SSL error:
https://github.com/NuGet/NuGetGallery/issues/7705
https://stackoverflow.com/questions/58993743/could-not-create-ssl-tls-secure-channel-while-trying-to-search-for-nuget-package
https://github.com/lukesampson/scoop/issues/2040
https://github.com/NuGet/NuGetGallery/issues/8176
https://github.com/chocolatey-community/chocolatey-test-environment/issues/48

**Comment by Be-ing** [comment]
At the end of the log you posted, when vcpkg starts to build ffmpeg, vcpkg used the copy of CMake 3.21.1 it downloaded to do that build:
```
[DEBUG] CreateProcessW("C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\vcpkg\downloads\tools\cmake-3.21.1-windows\cmake-3.21.1-windows-i386\bin\cmake.exe" "-DA
LL_FEATURES=all;all-gpl;all-nonfree;ass;avcodec;avdevice;avfilter;avformat;avisynthplus;avresample;bzip2;dav1d;fdk-aac;ffmpeg;ffplay;ffprobe;fontconfig;freetype
;fribidi;gpl;iconv;ilbc;lzma;modplug;mp3lame;nonfree;nvcodec;opencl;opengl;openh264;openjpeg;openssl;opus;postproc;sdl2;snappy;soxr;speex;ssh;swresample;swscale
;tensorflow;tesseract;theora;version3;vorbis;vpx;webp;x264;x265;xml2;zlib;" -DCURRENT_PORT_DIR=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/vcpkg/ports/ffmpe
g -D_HOST_TRIPLET=x64-windows "-DFEATURES=avcodec;avdevice;avfilter;avformat;core;swresample;swscale" -DPORT=ffmpeg -DVCPKG_USE_HEAD_VERSION=0 -D_VCPKG_DOWNLOAD
_TOOL=BUILT_IN -D_VCPKG_EDITABLE=0 -D_VCPKG_NO_DOWNLOADS=0 -DCMD=BUILD -DDOWNLOADS=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/vcpkg/downloads -DTARGET_TRIP
LET=x64-windows -DTARGET_TRIPLET_FILE=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/vcpkg/triplets/x64-windows.cmake -DVCPKG_BASE_VERSION=2021-09-10 -DVCPKG_C
ONCURRENCY=5 -DVCPKG_PLATFORM_TOOLSET=v142 -DGIT=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/vcpkg/downloads/tools/git-2.32.0.2-windows/mingw32/bin/git.exe
"-DVCPKG_PORT_CONFIGS=C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\build\vcpkg_installed\x64-windows\share\vcpkg-cmake\vcpkg-port-config.cmake;C:\Users\Joerg
\source\repos\JoergAtGithub\mixxx\build\vcpkg_installed\x64-windows\share\vcpkg-pkgconfig-get-modules\vcpkg-port-config.cmake" -DVCPKG_ROOT_DIR=C:/Users/Joerg/s
ource/repos/JoergAtGithub/mixxx/vcpkg -DPACKAGES_DIR=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/vcpkg/packages -DBUILDTREES_DIR=C:/Users/Joerg/source/repos
/JoergAtGithub/mixxx/vcpkg/buildtrees -D_VCPKG_INSTALLED_DIR=C:/Users/Joerg/source/repos/JoergAtGithub/mixxx/build/vcpkg_installed -DDOWNLOADS=C:/Users/Joerg/so
urce/repos/JoergAtGithub/mixxx/vcpkg/downloads -DVCPKG_MANIFEST_INSTALL=OFF -P "C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\vcpkg\scripts\ports.cmake")
```

GitHub Actions' runners [have CMake 3.21.3 currently](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md#tools). vcpkg should use CMake 3.21.1 that it downloads on GitHub Actions even though a newer CMake version is available.

**Comment by Be-ing** [comment]
I think you could hack around the CMake version mismatch by installing CMake 3.21.3 yourself.

From the links I posted, the SSL error seems to be an issue with your system configuration.

**Comment by Be-ing** [comment]
Please stop assuming random irrelevant things. If the cached packages didn't contain the expected data the build wouldn't work. There is no security issue. If the source code downloads didn't match the hash in vcpkg, it would fail. If any of the files in the vcpkg port were changed, the package and all its reverse dependencies would be rebuilt.

**Comment by daschuer** [comment]
I am not interested in chache hashes. I am interested in the origin of the files in the cache.
They have been created by a commit of this branch wich is no longer accessible.

**Comment by Be-ing** [comment]
I'm not wasting time on imagined problems. Digging through the build logs of every single dependency to look for link warnings is silly.

**Comment by daschuer** [comment]
The issue is unrelated to my local builds.

**Comment by daschuer** [comment]
Due to your latest rebase, all log files are no longer accessible. Was that intended? Why did you do this? Isn't this the opposite I have asked for? 

I am sure you still know that reabasing a PR under review must not be done without. :-(

**Comment by daschuer** [comment]
When you check the package revisions in GitHub: 

> Use GitHub Actions and GitHub's API to manage costs by automating package deletion.

> See online examples: https://github.com/marketplace/actions/delete-package-versions

**Comment by Be-ing** [comment]
I have verified that the macOS build artifact is still working.

@JoergAtGithub can you check that the Windows build artifact is still working? I do not expect any problems, but it is good to check with a major change to the build infrastructure.

**Comment by JoergAtGithub** [comment]
> 
> 
> From the links I posted, the SSL error seems to be an issue with your system configuration.

Indeed, I needed to add the following registry entries to enable TLS 1.2: https://github.com/NuGet/NuGetGallery/issues/8176#issuecomment-683923724 

Now I get:
```
    wavpack[core]:x64-windows -> 5.4.0
  * zlib[core]:x64-windows -> 1.2.11#13
  * zstd[core]:x64-windows -> 1.5.0
Additional packages (*) will be modified to complete this operation.
Restored 0 packages from C:\Users\Joerg\AppData\Local\vcpkg\archives in 2.347 ms. Use --debug to see more details.
Attempting to fetch 74 packages from nuget.
[DEBUG] CreateProcessW("C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\vcpkg\downloads\tools\nuget-5.10.0-windows\nuget.exe")
[DEBUG] cmd_execute_and_stream_data() returned 0 after   622000 us
[DEBUG] CreateProcessW("C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\vcpkg\downloads\tools\nuget-5.10.0-windows\nuget.exe" install "C:\Users\Joerg\source\rep
os\JoergAtGithub\mixxx\vcpkg\buildtrees\packages.config" -OutputDirectory "C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\vcpkg\packages" -Source mixxx-github-
packages -ExcludeVersion -PreRelease -PackageSaveMode nupkg -Verbosity detailed -ForceEnglishOutput -NonInteractive -DirectDownload -NoCache)
[DEBUG] cmd_execute_and_stream_data() returned 1 after  7319578 us
NuGet Version: 5.10.0.7240
Feeds used:
  https://nuget.pkg.github.com/mixxxdj/index.json

Restoring NuGet package fdk-aac_x64-windows.2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75.
Restoring NuGet package vcpkg-cmake-config_x64-windows.2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633cf99bbd.
Restoring NuGet package ffmpeg_x64-windows.4.4.0-vcpkge0200c2b1b5207929a52e05743648e9ebeff40900ff64d8f8be35c7aeec059f3.
Restoring NuGet package chromaprint_x64-windows.1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67.
  GET https://nuget.pkg.github.com/mixxxdj/download/fdk-aac_x64-windows/2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75/fdk-aac_x
64-windows.2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75.nupkg
  GET https://nuget.pkg.github.com/mixxxdj/download/vcpkg-cmake-config_x64-windows/2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633cf99b
bd/vcpkg-cmake-config_x64-windows.2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633cf99bbd.nupkg
  GET https://nuget.pkg.github.com/mixxxdj/download/chromaprint_x64-windows/1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67/chromapr
int_x64-windows.1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67.nupkg
  GET https://nuget.pkg.github.com/mixxxdj/download/ffmpeg_x64-windows/4.4.0-vcpkge0200c2b1b5207929a52e05743648e9ebeff40900ff64d8f8be35c7aeec059f3/ffmpeg_x64-wi
ndows.4.4.0-vcpkge0200c2b1b5207929a52e05743648e9ebeff40900ff64d8f8be35c7aeec059f3.nupkg
  NotFound https://nuget.pkg.github.com/mixxxdj/download/vcpkg-cmake-config_x64-windows/2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633
cf99bbd/vcpkg-cmake-config_x64-windows.2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633cf99bbd.nupkg 326ms
WARNING: Unable to find version '2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141633cf99bbd' of package 'vcpkg-cmake-config_x64-windows'.
  https://nuget.pkg.github.com/mixxxdj/index.json: Package 'vcpkg-cmake-config_x64-windows.2021.9.27-vcpkgb7ef4b796b6750d46aeae9c5363396baca73685f1d847800d74141
633cf99bbd' is not found on source 'https://nuget.pkg.github.com/mixxxdj/index.json'.

Restoring NuGet package fftw3_x64-windows.3.3.10-vcpkg3ec63d0a038552c8f541e64b838008ee73a73851ee77282959026804c9aec3da.
  GET https://nuget.pkg.github.com/mixxxdj/download/fftw3_x64-windows/3.3.10-vcpkg3ec63d0a038552c8f541e64b838008ee73a73851ee77282959026804c9aec3da/fftw3_x64-win
dows.3.3.10-vcpkg3ec63d0a038552c8f541e64b838008ee73a73851ee77282959026804c9aec3da.nupkg
  NotFound https://nuget.pkg.github.com/mixxxdj/download/chromaprint_x64-windows/1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67/chr
omaprint_x64-windows.1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67.nupkg 382ms
WARNING: Unable to find version '1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67' of package 'chromaprint_x64-windows'.
  https://nuget.pkg.github.com/mixxxdj/index.json: Package 'chromaprint_x64-windows.1.5.0-vcpkg098a7e56fcb0e76242dcdded1d22baa11ede2adf40f782b8051a038c5ea98d67'
 is not found on source 'https://nuget.pkg.github.com/mixxxdj/index.json'.

Restoring NuGet package hidapi_x64-windows.0.11.0-vcpkga604cfb5727d2f518da4697cc3b6e0242ecf4b8b7aaaa53ac35c054aa443e79c.
  GET https://nuget.pkg.github.com/mixxxdj/download/hidapi_x64-windows/0.11.0-vcpkga604cfb5727d2f518da4697cc3b6e0242ecf4b8b7aaaa53ac35c054aa443e79c/hidapi_x64-w
indows.0.11.0-vcpkga604cfb5727d2f518da4697cc3b6e0242ecf4b8b7aaaa53ac35c054aa443e79c.nupkg
  NotFound https://nuget.pkg.github.com/mixxxdj/download/fdk-aac_x64-windows/2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75/fdk-
aac_x64-windows.2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75.nupkg 435ms
WARNING: Unable to find version '2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75' of package 'fdk-aac_x64-windows'.
  https://nuget.pkg.github.com/mixxxdj/index.json: Package 'fdk-aac_x64-windows.2021.2.3-vcpkg137be6fdfad747b40b0fb071305d0204ed397678bce03884f797e3fba42b4a75'
is not found on source 'https://nuget.pkg.github.com/mixxxdj/index.json'.
```

**Comment by JoergAtGithub** [comment]
I don't like to modify my system for this. And it will not be a solution for others devs anyway, because the `x64 Native Tools Command Prompt for VS 2019` always uses the CMake version which comes with Visual Studio, which is in the moment 3.20.21032501-MSVC_2 . Don't know why GitHub Actions have a different version. Doesn't Github Actions use the described settings too?

**Comment by Be-ing** [comment]
I am puzzled why the macOS build is using CMake 3.21.1 but the Windows build is using CMake 3.21.3. Both the macOS and Windows GitHub Actions runners have CMake 3.31.3.

**Comment by Be-ing** [comment]
Hmmm I just checked the GitHub Packages for macOS and actually they are using CMake 3.21.3...

**Comment by Be-ing** [comment]
It is not individual packages that require specific CMake versions; vcpkg requires a specific CMake version for all packages in https://github.com/microsoft/vcpkg/blob/master/scripts/vcpkgTools.xml

The trouble is that vcpkg will use a newer version of CMake than the one in that XML file if the system has a newer version of CMake. On Windows, this is troublesome because vcpkg uses the CMake version from GitHub Actions.

On macOS, this is not such a problem because the version of CMake installed in the GitHub Actions runner comes from Homebrew:
https://github.com/actions/virtual-environments/blob/main/images/macos/provision/core/commonutils.sh
https://github.com/actions/virtual-environments/blob/main/images/macos/toolsets/toolset-12.json#L80

So I think for macOS, it would be easy to update the version of CMake used locally when GitHub Actions does because they both come from Homebrew.

None of this is great, but I think it is tolerable. A real solution would be for vcpkg to either [download its own copy of CMake unconditionally or not include the CMake version in the cache key](https://github.com/microsoft/vcpkg/issues/16615#issuecomment-947100263z).

**Comment by JoergAtGithub** [comment]
@Be-ing I appreciate your work on this very much! Thank you for this!
But to be honest I think VCPKG/NuGet binary caching is just not mature enough. And I fear that these issues could keep new contributors away from Mixxx. IMHO it's very important, that the very first build of a new contributor runs out of the box, but this process has too many hurdles to pass.
In general I like the concept, it's a much more elegant solution. But in practice, the old one runs much smoother.

**Comment by daschuer** [comment]
I have similar concerns. However it should work flawlessly on the CI. 
So I can imagine using it there, but default to the complete binary package download for home builds. 
What do you think?

**Comment by Holzhaus** [comment]
I think we should use the same system for both. Otherwise we might risk breaking the home builds without noticing.

**Comment by JoergAtGithub** [comment]
IMHO it's also a very good solution for users, who want to build the buildenv themself and therefore don't need binary caching. But I think this group is very small.

**Comment by Swiftb0y** [comment]
> C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\out\build\x64-Debug\vcpkg_installed\x64-windows\include\libusb-1.0\libusb.h(783): warning C4200: Nicht dem Standard entsprechende Erweiterung: Null-Array in Struktur/Union

"Non-standard extension: null-array in struct/union" ("null" might also mean zero here, not clear from the context)

>  C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\out\build\x64-Debug\vcpkg_installed\x64-windows\include\libusb-1.0\libusb.h(783): note: Dieser Member wird vom Standardkonstruktor oder Zuweisungsoperator für Kopieren/Verschieben ignoriert.

"This member will be ignored by the defaultconstructor or assignment operator for copying/moving."


Still weird that it fails in the libusb header and not mixxx. I guess its their responsibility to fix then?

**Comment by Swiftb0y** [comment]
Comment race-condition :sweat_smile:

**Comment by JoergAtGithub** [comment]
BTW: This kind of issues could be prevented by CMake presets, because than local builds will use the same presets as CI.

**Comment by Be-ing** [comment]
Please attach C:\Users\Joerg\source\repos\JoergAtGithub\mixxx\out\build\x64-Debug\vcpkg-manifest-install.log

**Comment by JoergAtGithub** [comment]
I deleted mixxx\vcpkg\packages and could rebuild. With BULK==off, the build succeeded!

**Comment by Be-ing** [comment]
@JoergAtGithub how exactly could we explain how to install the English language pack in BUILDING.md?

**Comment by Be-ing** [comment]
With https://github.com/microsoft/vcpkg/pull/21471 and the new `--x-abi-tools-use-exact-versions` option, I think the issue with cache misses on local builds due to different CMake and PowerShell versions is finally fixed. @JoergAtGithub please test building this branch again to confirm that vcpkg hits the cache. 

Once https://github.com/mixxxdj/vcpkg/pull/33 is merged I will switch the vcpkg submodule to the mixxxdj/2.4 branch and I think we can finally merge this.

**Comment by Be-ing** [comment]
I confirm the cache miss issues are solved now. I built locally on macOS and vcpkg hit the cache. Then I upgraded CMake with Homebrew, deleted the CMake build directory, and vcpkg still hit the cache.

File: BUILDING.md
Comment (ywwg) [reply BUILDING.md:1]: usually a file like this is called COMPILING.md but this is a nitpick

**Review by ywwg — COMMENTED** [review]
I think the submodule method is the right way to go. This PR contains a lot of changes to a lot of platforms though so it needs pretty extensive hand-testing.

**Comment by Be-ing** [comment]
I took another look at how Tenacity is set up to automatically configure read-only access to GitHub Packages. The trick is pretty simple; just split the GitHub personal access token into two strings so GitHub doesn't automatically deactivate it. This allows for simplifying the documentation; there is no need to run the `nuget` command manually.

**Comment by Be-ing** [comment]
I have tested this on Windows and I happy to report that it Just Works, both the command line instructions in BUILDING.md and loading the repository in Visual Studio (VSCode works too, but the CMake generator needs to specifically be set to Ninja). Aside from installing the required tools, there are no longer any extra steps on any OS; CMake and vcpkg take care of everything. This is the first time I've been able to compile Mixxx locally on Windows. I do not think we will have Windows developers failing to get started because of confusion running the old batch script. Plus, this makes Mixxx buildable on ARM macOS (as an x86-64 executable).

All that is left to do is merge https://github.com/mixxxdj/vcpkg/pull/33 then switch the vcpkg submodule on this branch to the mixxxdj/2.4 branch.

**Comment by Be-ing** [comment]
I have extracted some generally useful code to initialize vcpkg into a CMake module and opened a PR to share it upstream: https://github.com/microsoft/vcpkg/pull/21591

**Pull request closed without merging**
